harvester/harvester
# supportbundle: report progress

Dependency: https://github.com/harvester/support-bundle-utils/pull/6

**Problem:**
Need to provide progress of support bundle creation.

**Solution:**
- Do not update CR state in support bundle manager. Instead, the support
  bundle manager export bundle collecting status.
- The controller update state and progress in the reconciliation loop.
- Add a timeout: in some situations the support bundle manager pod can't
  never be created, adding a timeout to prevent endless wait.
- Move shared types to a package thus the support bundle manager code
  can use it.

**Related Issue:**
https://github.com/harvester/harvester/issues/783

**Test plan:**


**Test by creating a supportbundle resource**

- Wait for https://github.com/harvester/support-bundle-utils/pull/6 to be merged. Or use the latest dev image for that PR:
   ```
   kubectl patch setting support-bundle-image --type json -p '[{"op": "replace", "path": "/value", "value":"bk201z/support-bundle-utils:pr6"}]'
   ```
-  Edit the setting `support-bundle-image-pull-policy` to `Always`
   ```
   kubectl patch setting support-bundle-image-pull-policy --type json -p '[{"op": "replace", "path": "/value", "value":"Always"}]'
   ```

- Patch a running harvester backend with the image of the PR

  ```
  kubectl patch deployment harvester -n harvester-system --type json -p '[{"op": "replace", "path": "/spec/template/spec/containers/0/image", "value":"bk201z/harvester:pr844"}]'
  ```

- Create a supportbundle

  ```
  $ cat sb.yaml
  ---
  apiVersion: harvesterhci.io/v1beta1
  kind: SupportBundle
  metadata:
    name: foo
    namespace: harvester-system
  spec:
  description: foo
  
  $ kubectl create -f sb.yaml
  supportbundle.harvesterhci.io/foo created
  ```


- The `progress` field should report the progress. After some seconds, the bundle state should be ready.
  ```
  $ kubectl get sb foo -n harvester-system -o yaml
  <...>
    issueURL: ""
  status:
    progress: 40
    state: generating

  $ kubectl get sb foo -n harvester-system -o yaml
  <...>
  status:
  conditions:
  - lastUpdateTime: "2021-05-27T06:30:20Z"
    status: "True"
    type: Initialized
  filename: harvester-supportbundle_27b74135-b255-4916-bb7d-b0186444a92b_2021-05-27T06-29-58Z.zip
  filesize: 1848274
  progress: 100
  state: ready
  ```

**Test in UI**

Current UI should also work with the change. Follow the steps in privous section to patch the image and click the `Generate Support Bundle` link at the bottom of UI to generate and download a support bundle.
